README.md
<h3 align="center">
  <img src="ReadmeData/AppDeployLogo.png" alt="AppDeploy Logo" />
</h3>

Visualize, template and deploy your app in a second.
============

AppDeploy is the fastest way to get info and deploy your mobile app Over The Air without specific server configuration (it only uses html).

## Features
- [x] ✨ Get quick info on your iOS and Android App (support of .xcarchive, .ipa and .apk)
- [x] 🗃 iOS: Get fast access to your app resources, plist, embbeded profiles…
- [x] 🚨 iOS: Get notified when an xcarchive is built
- [x] 🌅 Use default template (branded or with app icon) or customize your own one
- [x] 🚀 Deploy to your server without extra configuration just html support is required
- [x] 🎉 Notify deployment success with Hipchat or Prowl
- [x] 🚠 Integrate it in your Continuous Integration by using the command line
- [x] 📲 Download the app with a versioned url

## Current usage
- Currently used in our production toolchain with a Jenkins building the app with gym (fastlane) and AppDeploy is generating the download page in script mode from the .ipa or .apk file.
- It is also nice to have info quickly on App by double clicking the files (requires to define AppDeploy as the default App to open your ipa/apk/xarchive files).

## Download and installation
You can build or download a ready to use version available and stay automatically up to date thanks to Sparkle. [Download here](https://www.atelierdumobile.com/apps/releases/appdeploy)

## Requirements
- A mac with OSX10.10

## Presentation

#### Supported files
Drag any apk, xcarchive, ipa

![Gifs](ReadmeData/SupportedFiles.gif)


#### Choose the template you like
You have default template and you can create the one that fits your needs.
![Gifs](ReadmeData/AnyTemplateYouWant.gif)


#### Deploy in a click
Choose your template and network config and deploy.

![Gifs](ReadmeData/ipa.gif)

![Gifs](ReadmeData/apk.gif)

#### Full integration with Xcode archive
Archive -> AppDeploy detects the archive and allow you to handle it.
![Gifs](ReadmeData/FullWorkflow.gif)


#### Rich settings and options
Discovers the settings and discovers the settings and possible customization.
![Png](ReadmeData/SettingsTerminal.png)


## Possible Improvements (feedback appreciated)

Display

	- Display more information such as size of resources and repartition
	- Support Mac application
	- Nicer display among supported plateform
	- MultiWindow support
	- Add a comment section

Network

	- Improvement of the network upload reliability
	- SSH mode has no progress bar
	- Dropbox support for sharing
	- Generate random url for privacy

Templating

	- Improvement of the template management
	- Allow a template store to share its template

Uploading

	- Cancel a build/upload in progress

Settings


Notification

	- Slack support

Knowns bugs or assimilated

	- ⚠️ Store SFTP password in keychain or encrypt them in the settings (ssh mode do not store pwd so is currently secured)  

	- clean code, put a CI & and unit test back

## Feature requests and feedback
Ping me on [twitter](http://twitter.com/nlauquin)

If you want to help me, tell me which usage you will use or like to do with this tool:
- 1/ Easy way to consult app technical details
- 2/ Integration in your toolchain (Jenkins)
- 3/ Manual usage for templating and deploy app
- Which part to remove/which to focus on?


## License

This project is licensed under the terms of the BSD license. See the [LICENSE](LICENSE) file.


AppDeploy/ChangeLog.md
AppBuilder
==========

Allow to deploy beta from xcarchive Xcode build-in system and apk basic support for Android.

ChangeLog
==========
#4.5/54
- Compatibility with Xcode 9
- In the setting, improve the SFTP test by checking the presence of the root path
- Disable the xcharchive build feature

#4.4/53
- New App icons

#4.3/52
- Integrate Fabric Crashlytics and allow app to crash (?? :p)
- Add keyboard shortcuts & easier keyboard navigation between textfield

#4.2/51
- Fix bug about the script section
- Rename default template file
- Add key for template & server to allow easy usage on script
- Minor bug fixes and text corrections

#4.1/50
- Fix Android issue concerning the detection of the app label
- Android: support for architecture display and locales display in about more information

#4.0/49
- IPA displays provisionning profile & signing information
- Change name to AppDeploy
- Compatibility starting OSX 10.10

#3.1/48
- Distinguish sound success to sound error
- Several bug fix
- Configuration with no network by default
- Settings : new android section, add terminal section, add help for network, documentaiton template
- Update templates tags

#3.0.4/46
- Android : detect ic_launcher whern stored in mipmap folder

#3.0.2/44
- Commun: fix export encoding issue that prevent to generate file.

#3.0.1/43
- Android: improve icone selection for better quality
- Android: fix issue with packagename parsing when _ in the name

#3.0.0/41
- Android now use Aapt to get information about the apk to distribute (taking the bundleid, version, sdk…)
- Adding new value template [[OS_MIN]], [[SDK]], [[BUNDLE_IDENTIFIER]]
- Aapt file is now configurable through the configuration file (and is readonly through the pref yet) 

#2.9.9/40
- Internal Android icone usage 

#2.9.8/39
- Fix url to app binary provided in output command line generation
- export : export the bundleidentifier

#2.9.8/38
- Adding --export=filepath.properties to export some property info about the package
- Template can fill the uniqueintegration number with [[IC_UNIQUE_NUMBER]] value

#2.9.7/37
- IC : using --build_number=20 parameter will use has a subfolder the argument provided instead of the technical build number 
- add integration for support ADM_IC

#2.9.6/36
- update UI to clarity and support network config local
- fix & update command line support to use template and network settings

#2.9.5/35
- update command line support to use configuration

#2.9/34
- Fix ipa detection support

#2.9/33
- Fix template item preference
- Reorganize menu
- Add item "Generate template without binary"
- Add logo url for template

#2.9/32
- Sparkel auto-update

#2.9/31
- Fix issue on 10.9
- Block add new network in adm releases

#2.9/30
- Improve menus && enable in context
- Add shortcut when pressing "cmd" allows to resend the build without regenerating the archive
- Network: support ssh passwordless but no progress bar in such cases
- Network setting : support duplicate label name
- Network : transfert return a error code
- Network & batch : cancelable tasks

#2.8/29
- Change xcarchive build method to support entitlements
- Default ADM config adding
- Use external configuration with global settings including (except the integration services)
- Add icone to app
- Ipa Icone support
- Display Entitlements for ipa & xcarchive

#2.7/28
- New settings to handle network configuration
- New settings to handle template & template preview

#2.6/27
- Restore support for beta & release
- Allow to define a custom output directory
- HipChat support

#2.5/26
- New template with app icone

#2.3.7/25
- Adding a template section in settings
- Date format choosable in the template preference.

#v2.3.6/24
- Display date
- Bugfix for data display when switching app
- Adding fast open url, copy board & notification with shortcuts
- Adding open of mobile provisionning profile
- Xcode setting
- UI enlarge window width

#v2.3.5/22
- adding hardcoded setting and disable settings
- android basic ui fix

#v2.3.0/21
- url is now with the technical version url and not the functionnal version url

#v2.2.0/20
- size of file
- improved settings with credentials

#v2.1.0/18
- support basique CLI
- support ipa

#v2.0.3/17
- drag and drop supported for xcarchive & apk file
- display file source in title

#v2.0.2/16
- wording update
- fix a problem with prowl due to Android implementation
- the "open build folder" button is available as soon as the build is signed (vs when uploaded)
- add extra information when click on the app icone: (sdk, minimum version, app store size)

#v2.0.1/15
- display the CFBundleVersion in the html page (which is the technical version unique identifier of a build)

#v2.0/12
- android support & ios adjustments

#v.9.5/10
- Code refactor
- Name is normalized in lowercase with only normal caracter (letter & number _-)
- Settings for opening or not build folder automatically after a deployment
- Local notification is done at the end of a deployement

#v0.9
- Update preference to support :
 - Can open temporary folder for cleaning
- Automatic last archive check at startup

#v0.8.5
- You can disable prowl notification
- Logs are available in the console
- Remove a certification sigature error which is not blocking anymore on OSX10.9.5

#v0.8
- Network : add a progress status for upload
- Network : catch network errors

#v0.7
- Add on home "last archive detection"
- Fix back button bad state

#v0.6
- Add prowl support
- Add bip at the end of the build

#v0.5
- Add asynchrone task and success status

#v0.3
- Fix for iOS7.1 SSL issue
- Sign the app using the embedded provisionning profile


AppDeploy/AppDeploy/Constants.h
//For development settings
#ifdef DEBUG
	#define kEnableProwl				YES
	#define kHipChatEnabled				YES
	#define kCleanTempFolderAtStartup	NO
#else
	#define kEnableProwl				YES
	#define kHipChatEnabled				YES
	#define kCleanTempFolderAtStartup	YES
#endif
#define kPlaySound YES

//Configuration
#define kConfigFolder 					@"com.atelierdumobile.AppDeploy"
#define kConfigJsonFile 				@"config.json"

//Temporary Folders
#define kFolderPrefixTempZipExtract 	@"TEMP_ZIP_Extract"
#define kMainTemporaryFolder 			@"/APPDEPLOY_TEMP/"
#define kTemporaryPreviewSubFolder 		@"AppDeployPreview"
#define TEMP_FOLDER_PREFIX 				@"BUILD_"

//File detection latency
#define ARCHIVE_LATENCY					((NSTimeInterval)4.0)

//Terminal
#define kTerminalPath 					@"/Applications/Utilities/Terminal.app"
#define kCommandLineHelpOutput 			@"/Contents/MacOS/AppDeploy --help"
#define SHELL_SUCCESS_CODE 				0
#define SHELL_FAILURE_CODE				1
#define kICConsantPrefix				@"appdeploy_"

//IOS
#define kPlayloadFolder 				@"Payload/"
#define kEmbeddedProvisionningFileName 	@"embedded.mobileprovision"
#define kArchiveFolderPath 				@"/Library/Developer/Xcode/Archives"
#define kDefaultProvisioningName 		@""
#define kIOSPlateforme 					@"ios"
#define kIOSXCArchiveExtension 			@".xcarchive"
#define kIOSIPAExtension 				@".ipa"

//Android
#define ANDROID_LAUNCHER_FILE_NAME 		@"ic_launcher.png"
#define kAaptToolName 					@"aapt"
#define kPlistFileName  				@"Info.plist"
#define KProductFolder 					@"Products"
#define kAndroidPlateforme 				@"Android"
#define kAndroidPackageExtension 		@".apk"


//Template
#define kAndroidSuffix 					@"_android"
#define kIOSSuffix 						@"_ios"

//Template 1
#define kDefaultTemplateModel1File 		@"defaultCompany.html"
#define kDefaultTemplateModel1Label		@"🔒 Builtin - logo company"
#define kDefaultTemplateModel1Logo		@"http://atelierdumobile.com/images/companyLogo.png"
#define kDefaultTemplateModel1Key		@"defaultCompany"

#define kDefaultTemplateModel2File 		@"defaultAppIcone.html"
#define kDefaultTemplateModel2Label		@"🔒 Builtin - app icone"
#define kDefaultTemplateModel2Logo		@""
#define kDefaultTemplateModel2Key		@"defaultAppIcone"


//Network
#define kNotificationUploadProgress 	@"NotificationUploadProgress"
//#define kTimeoutConnection 15

//Scripting
#define kScriptDefaultConfig 			kDefaultTemplateModel1Key
#define kScriptOpenTemplateAppFolder	YES
